boot/bootutil/src/loader.c

```diff
@@ -1551,89 +1558,238 @@ boot_set_image_ok(void)
 #endif /* !MCUBOOT_OVERWRITE_ONLY */
 
 /**
- * Performs an image swap if one is required.
+ * Sets the active image on which the bootloader will continue to work.
+ *
+ * @param image_id                  The index of the image (from 0).
+ */
+static void
+boot_set_active_image(size_t image_id)
+{
+    current_image = image_id;
+    boot_data = &boot_data_array[image_id];
+
+#ifdef MCUBOOT_ENC_IMAGES
+    /* The keys used for encryption are no longer valid (may belong to
+     * another image).
+     */
+    boot_enc_zeroize();
+#endif
+
+    return;
```

Comment:
If the secondary slot (slot1) is encrypted the keys are stored in an enc_key_data structure which will be marked as "valid" after the keys have been set (the keys won't be reloaded because they are "valid"). When the MCUBoot switches between images the keys must be reloaded.; therefore, it clears the encryption keys and the valid flag too.

Reply:
Hehe, I was way to ambiguous in my comment. I was wondering why the empty return statement was there.

Reply:
I will remove it.